src/instructions.rs

```diff
@@ -0,0 +1,814 @@
+pub fn add(instr: u16, regs: &mut Registers) -> Result<(), VMError> {
+    // Destination register
+    let dr: Register = Register::from_u16((instr >> 9) & THREE_BIT_MASK)?;
+    // First operand
+    let sr1: Register = Register::from_u16((instr >> 6) & THREE_BIT_MASK)?;
+    // Check the bit 5 to see if we are in immediate mode
+    let imm_flag = (instr >> 5) & ONE_BIT_MASK;
+
+    if imm_flag == 1 {
+        // Get the 5 bits of the imm5 section (first 5 bits) and sign extend them
+        let mut imm5 = instr & FIVE_BIT_MASK;
+        imm5 = sign_extend(imm5, 5)?;
+        regs[dr] = regs[sr1].wrapping_add(imm5);
+    } else {
+        // Since the immediate flag was off, we only need the SR2 section (first 3 bits).
+        // This section contains the register containing the value to add.
+        let sr2 = Register::from_u16(instr & THREE_BIT_MASK)?;
+        regs[dr] = regs[sr1].wrapping_add(regs[sr2]);
+    }
+
+    update_flags(dr, regs);
+    Ok(())
+}
+
+/// Does the bitwise 'NOT' for a value stored in a register.
+///
+/// ### Arguments
+///
+/// - `instr`: An u16 that has the encoding of the whole instruction to execute.
+/// - `regs`: A Registers struct that handles each register.
+pub fn not(instr: u16, regs: &mut Registers) -> Result<(), VMError> {
+    let dr = Register::from_u16((instr >> 9) & THREE_BIT_MASK)?;
+    let sr = Register::from_u16((instr >> 6) & THREE_BIT_MASK)?;
+
+    regs[dr] = !regs[sr];
+    update_flags(dr, regs);
+    Ok(())
+}
+
+/// Does the bitwise 'AND' between two values and stores the result
+/// in a register.
+///
+/// This instruction can happen in two different ways. One is
+/// by doing it with two elements that are in a register each,
+/// this is called register mode. The other mode is to use the value
+/// of a register with the one that is embedded in the instruction itself.
+///
+/// ### Arguments
+///
+/// - `instr`: An u16 that has the encoding of the whole instruction to execute.
+/// - `regs`: A Registers struct that handles each register.
+pub fn and(instr: u16, regs: &mut Registers) -> Result<(), VMError> {
+    // Destination register
+    let dr = Register::from_u16((instr >> 9) & THREE_BIT_MASK)?;
+    // SR1 section
+    let sr1 = Register::from_u16((instr >> 6) & THREE_BIT_MASK)?;
+    // Imm flag
+    let imm_flag = (instr >> 5) & ONE_BIT_MASK;
+
+    if imm_flag == 1 {
+        // Get the imm5 section, then do the bitwise and with the content on R1.
+        let mut imm5 = instr & FIVE_BIT_MASK;
+        imm5 = sign_extend(imm5, 5)?;
+        regs[dr] = regs[sr1] & imm5;
+    } else {
+        // Get the SR2 section, then do the bitwise and with the content on R1.
+        let sr2 = Register::from_u16(instr & THREE_BIT_MASK)?;
+        regs[dr] = regs[sr1] & regs[sr2];
+    }
+
+    update_flags(dr, regs);
+    Ok(())
+}
+
+/// Changes the PC register if the Cond register is set to the
+/// flag that is selected on the encoding of the instruction
```

Comment:
I would rephrase this. I'm thinking of separating this into two sentences, for example:

```suggestion
/// Changes the PC register value depending on the value of the Cond register.
/// If the Cond register is set to the same value taken from the bits 9 to 11 of the instruction,
/// this function adds the PC register to the value taken from the first 9 bits of the instruction. 
```